packages/misc/src/lib/helper/session-sigs-validator.ts

```diff
@@ -0,0 +1,208 @@
+  // }
+  return parsedData;
+}
+
+// Function to validate expiration date
+function validateExpiration(
+  expirationTimeStr: string,
+  context: string
+): ValidationResult {
+  const errors: string[] = [];
+  const expirationTime = new Date(expirationTimeStr);
+  const currentTime = new Date();
+
+  if (isNaN(expirationTime.getTime())) {
+    errors.push(
+      `Invalid Expiration Time format in ${context}: ${expirationTimeStr}`
+    );
+  } else if (expirationTime < currentTime) {
+    errors.push(
+      `Expired ${context}. Expiration Time: ${expirationTime.toISOString()}`
+    );
+  }
+
+  return {
+    isValid: errors.length === 0,
+    errors,
+  };
+}
+
+// Function to parse and validate capabilities
+function parseCapabilities(capabilities: Capability[]): ValidationResult {
+  const errors: string[] = [];
+
+  capabilities.forEach((capability, index) => {
+    const { signedMessage } = capability;
+
+    // Parse the signedMessage
+    const parsedCapabilityMessage = parseSignedMessage(signedMessage);
+    capability.parsedSignedMessage = parsedCapabilityMessage;
+
+    // Extract and validate expiration date
+    const expirationTimeStr = parsedCapabilityMessage['Expiration Time'];
+
+    if (expirationTimeStr) {
+      const validationResult = validateExpiration(
+        expirationTimeStr,
+        `capability ${index}`
+      );
+      if (!validationResult.isValid) {
+        errors.push(...validationResult.errors);
+      }
+    } else {
+      errors.push(
+        `Expiration Time not found in capability ${index}'s signedMessage.`
+      );
+    }
+  });
+
+  return {
+    isValid: errors.length === 0,
+    errors,
+  };
+}
+
+/**
+ * Validates the session signature.
+ *
+ * @param sessionSig - The session signature to validate.
+ * @returns The validation result, indicating whether the session signature is valid and any errors encountered during validation.
+ */
+export function validateSessionSig(sessionSig: AuthSig): ValidationResult {
+  const errors: string[] = [];
+
+  // Parse the main signedMessage
+  let parsedSignedMessage: ParsedSessionMessage;
+  try {
+    parsedSignedMessage = JSON.parse(sessionSig.signedMessage);
+  } catch (error) {
+    errors.push('Main signedMessage is not valid JSON.');
+    return { isValid: false, errors };
+  }
+
+  // Validate capabilities
+  const capabilities: Capability[] = parsedSignedMessage.capabilities;
+
+  if (!capabilities) {
+    errors.push('Capabilities not found in main signedMessage.');
+  } else if (capabilities.length === 0) {
+    errors.push('No capabilities found in main signedMessage.');
+  } else {
+    const capabilitiesValidationResult = parseCapabilities(capabilities);
+
+    if (!capabilitiesValidationResult.isValid) {
+      errors.push(...capabilitiesValidationResult.errors);
+    }
+  }
+
+  // Validate outer expiration
+  const outerExpirationTimeStr = parsedSignedMessage['expiration'];
```

Comment:
Shouldn't `expiration` be defined in `ParsedSignedMessage`?

Reply:
no - because `expiration` is part of the authSig/one of the sessionSigs. However, `ParsedSignedMessage` refers to the string property `signedMessage`, and we need to parse this string into JSON in order to extract the `expiration` property.

Reply:
You are right! the inner/out authSig's structure confused me!